shared/lib/accounts/accounts.ts

```diff
@@ -35,46 +50,188 @@ export function getUniqueAccountName(
+  getNextAvailableAccountName: () => Promise<string>,
+  snapId: SUPPORTED_WALLET_SNAP_ID,
+  { chainId }: SnapAccountNameOptions = {},
 ) {
-  const snapId = SOLANA_WALLET_SNAP_ID;
-  const scopes = [SolScope.Mainnet, SolScope.Testnet, SolScope.Devnet];
-  const client = new KeyringInternalSnapClient({
-    messenger: controllerMessenger,
-    snapId,
-  });
-
-  for (let index = 0; ; index++) {
-    const discovered = await client.discoverAccounts(
-      scopes,
-      entropySource,
-      index,
-    );
+  const defaultSnapAccountName = await getNextAvailableAccountName();
+
+  // FIXME: This is a temporary workaround to suggest a different account name for a first party snap.
+  const accountNumber = defaultSnapAccountName.trim().split(' ').pop();
 
-    // We stop discovering accounts if none got discovered for that index.
-    if (discovered.length === 0) {
-      break;
+  switch (snapId) {
+    case BITCOIN_WALLET_SNAP_ID: {
+      if (chainId === MultichainNetworks.BITCOIN_TESTNET) {
+        return `Bitcoin Testnet Account ${accountNumber}`;
+      }
+      return `Bitcoin Account ${accountNumber}`;
+    }
+    case SOLANA_WALLET_SNAP_ID: {
+      // Solana accounts should have in their scope the 3 networks
+      // mainnet, testnet, and devnet. Therefore, we can use this name
+      // for all 3 networks.
+      return `Solana Account ${accountNumber}`;
     }
+    default:
+      return defaultSnapAccountName;
+  }
+}
+
+///: BEGIN:ONLY_INCLUDE_IF(solana,bitcoin)
+export type WalletSnapOptions = {
+  scope?: CaipChainId;
+  derivationPath?: DiscoveredAccount['derivationPath'];
+  entropySource?: string;
+  accountNameSuggestion?: string;
+};
+
+export type WalletSnapClient = {
+  getSnapId(): SnapId;
+
+  createAccount(
+    options: WalletSnapOptions,
+    internalOptions?: SnapKeyringInternalOptions,
+  ): Promise<KeyringAccount>;
+
+  getNextAvailableAccountName(
+    options?: SnapAccountNameOptions,
+  ): Promise<string>;
+};
+
+type MultichainWalletSnapClientMessenger = Messenger<
+  | SnapControllerHandleRequest
+  | AccountsControllerGetNextAvailableAccountNameAction,
+  never
+>;
+
+export class MultichainWalletSnapClient implements WalletSnapClient {
+  #snapId: SUPPORTED_WALLET_SNAP_ID;
+
+  #snapKeyring: SnapKeyring;
+
+  #scopes: CaipChainId[];
+
+  #client: KeyringInternalSnapClient;
 
-    await Promise.allSettled(
-      discovered.map(async (discoveredAccount) => {
-        const options = {
-          derivationPath: discoveredAccount.derivationPath,
-          entropySource,
-        };
-
-        // TODO: Use `withKeyring` instead of using the keyring directly.
-        await snapKeyring.createAccount(snapId, options, {
-          displayConfirmation: false,
-          displayAccountNameSuggestion: false,
-          setSelectedAccount: false,
-        });
+  #messenger: MultichainWalletSnapClientMessenger;
+
+  constructor(
+    snapId: SUPPORTED_WALLET_SNAP_ID,
+    scopes: CaipChainId[],
```

Comment:
This does seem a bit weird to not couple `snapId` and their matching `scopes`. A better implementation could be to have a `{Solana,Bitcoin}WalletSnapClient` where we would move the specific logic on each clients (like the account naming for example).